cli/command/swarm/progress/root_rotation.go

```diff
@@ -0,0 +1,138 @@
+package progress
+
+import (
+	"bytes"
+	"io"
+	"os"
+	"os/signal"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/swarm"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/progress"
+	"github.com/docker/docker/pkg/streamformatter"
+	"github.com/opencontainers/go-digest"
+)
+
+const (
+	certsRotatedStr = "  rotated TLS certificates"
+	rootsRotatedStr = "  rotated CA certificates"
+	// rootsAction has a single space because rootsRotatedStr is one character shorter than certsRotatedStr.
+	// This makes sure the progress bar are aligned.
+	certsAction = ""
+	rootsAction = " "
+)
+
+// RootRotationProgress outputs progress information for convergence of a root rotation.
+func RootRotationProgress(ctx context.Context, dclient client.APIClient, progressWriter io.WriteCloser) error {
+	defer progressWriter.Close()
+
+	progressOut := streamformatter.NewJSONProgressOutput(progressWriter, false)
+
+	sigint := make(chan os.Signal, 1)
+	signal.Notify(sigint, os.Interrupt)
+	defer signal.Stop(sigint)
+
+	var updater *rootRotationProgressUpdater
+
+	for {
+		info, err := dclient.SwarmInspect(ctx)
+		if err != nil {
+			return err
+		}
+
+		if updater == nil {
+			updater = &rootRotationProgressUpdater{
+				progressOut: progressOut,
+			}
+		}
```

Comment:
The lazy initialization doesn't seem necessary. Normally I'd say to just define an updater outside the for loop. However, I'm not sure the abstraction is justified here. In service create/update, there are two implementations. Here, there is only one implementation, so it seems like it could just be a function call `updateProgress` or something.`done` should become a return value. As for `initialized`, you could just move those initial `progress.Update` calls to happen before the for loop in this function.